README.rst
A Bulk-Upload Library for DocumentCloud
---------------------------------------

pneumatic is a Python 3 library that adds some luxury and safeguards to the bulk-uploading of hundreds, thousands or hundreds of thousands of files to `DocumentCloud <https://www.documentcloud.org>`_. It is meant to do one thing -- upload -- and serve as an adjunct to, but not a replacement for, the excellent `python-documentcloud <http://python-documentcloud.readthedocs.io/en/latest/>`_ API wrapper.

pneumatic's name is inspired by the pneumatic dispatch systems in newsrooms of yore, which featured a `series of pneumatic tubes`_ for sending copy from the newsrooms to other departments such as the composing room.

Features
--------

* Catalogs the API response for each upload in a SQLite database along with the file's canonical URL.
* Post-processing, can update the SQLite database with each document's page count and file hash.
* Dumps the SQLite data to a CSV if you wish.
* Prevents inadvertent submission of file types DocumentCloud doesn't handle, such as audio.

Links
-----

* Documentation:    https://pneumatic.readthedocs.io/en/latest/
* Repository:       https://github.com/anthonydb/pneumatic
* Issues:           https://github.com/anthonydb/pneumatic/issues

Basic Usage
-----------

You will need an active DocumentCloud account and Python 3.5+. First, install via pip:

.. code-block:: python

    pip install pneumatic

Example use: To upload all files in a directory (and all sub-directories below it), assign them to an existing project, set the files to public access, and tag each with metadata, run the following code:

.. code-block:: python

    from pneumatic import DocumentCloudUploader

    uploader = DocumentCloudUploader('person@example.com', 'your-password')
    uploader.upload(
        file_directory='/govfiles',
        project='17477-loudoun-county-government',
        access='public',
        data={'type': 'government', 'action': 'lawsuit'})

Please see the `full documentation`_ for more examples, including how to access the uploads database.

.. _`series of pneumatic tubes`: https://en.wikipedia.org/wiki/Pneumatic_tube
.. _`full documentation`: https://pneumatic.readthedocs.io/en/latest/


CHANGELOG.rst
Changelog
=========

0.1.9 - January 18, 2020
--------------------------

* Fix issue with authenticating API credentials via Requests. (`#20 <https://github.com/anthonydb/pneumatic/issues/20>`_)
* Disable multiprocessing for now. (`#21 <https://github.com/anthonydb/pneumatic/issues/21>`_)
* Confirm Python 3.8 support.

0.1.8 - September 23, 2018
--------------------------

* Improve command-line message content and formatting.
* Confirm Python 3.6 and 3.7 support.
* Expand number of unsupported filetypes to exclude from import list.
* Improve CSV output file name formatting.
* Documentation edits and updates.

0.1.7 - November 24, 2016
-------------------------

* Delete SQLite database if, upon program exit, it contains no records. (`#14 <https://github.com/anthonydb/pneumatic/issues/14>`_)
* Add some color, improve readability in console output. (`#18 <https://github.com/anthonydb/pneumatic/issues/18>`_)
* Add colorama dependency for printing ANSI escape code colors in Windows.

0.1.6 - April 28, 2016
----------------------

* Provide ``User-Agent`` and ``From`` HTTP header fields.
* Correct issue where ``force_ocr`` and ``secure`` parameters were not being set default to ``false`` in the Ruby way. (`#17 <https://github.com/anthonydb/pneumatic/issues/17>`_)
* Handle 50X errors from the API (which do not return JSON). (`#15 <https://github.com/anthonydb/pneumatic/issues/15>`_, thank you, `Tom Meagher <https://github.com/tommeagher>`_!)

0.1.5 - March 9, 2016
---------------------

* Add ``update_processed_files`` method to get page, file hash and other data that's not available upon upload.
* Add ``pages``, ``file_hash``, ``id`` and ``title`` to items tracked in database.
* Create database upon ``DocumentCloudUploader`` initialization.
* Get rid of the file extension in the document title. (`#13 <https://github.com/anthonydb/pneumatic/issues/13>`_)
* Bug fix: Properly test for presence of data and title keyword arguments.

0.1.4 - February 17, 2016
-------------------------

* Remove extra line space on csv dump in Windows.

0.1.3 - February 16, 2016
-------------------------

* Report number of files to be uploaded before starting.
* Better reporting of upload progress and results.
* More comprehensive filetype exclusion list.
* Record pdf and text URLs in database.
* Uploads that return status codes other than 200 are handled.

0.1.2 - February 4, 2016
------------------------

* ``dump_to_csv`` outputs contents of SQLite database. (`#2 <https://github.com/anthonydb/pneumatic/issues/2>`_)
* Add ``force_ocr`` parameter to upload options.
* Removed multiprocessing support for Windows for now.
* Report when upload file directory does not exist.
* Better testing for prohibited file types.
* Only create database after file path verified.

0.1.1 - December 31, 2015
-------------------------

* Packaged for release to PyPi.

0.1 - December 16, 2015
-----------------------

* Add multiprocessing. (`#1 <https://github.com/anthonydb/pneumatic/issues/1>`_)
* Exclude files of 400MB or larger from upload. (`#3 <https://github.com/anthonydb/pneumatic/issues/3>`_)
* Add initial tests.
* Scaffolding for documentation.

0.0.1 - November 26, 2015
-------------------------

* Pre-alpha prototype


tests/test_utils.py
#!/usr/bin/env python

import unittest
import tempfile
from pneumatic.utils import *


class TestFileSanitize(unittest.TestCase):

    def test_flag_file_type(self):
        """
        sanitize_uploads() should flag a prohibited file type
        with True in the returned exclude_flag field.
        """
        self.utils = Utils()
        files_dict_list = [{'name': 'SomeFile.mp3',
                            'full_path': '/path/SomeFile.mp3',
                            'exclude_flag': None,
                            'exclude_reason': None
                            }]
        docs = self.utils.sanitize_uploads(files_dict_list)
        self.assertTrue(docs[0]['exclude_flag'])

    def test_flag_file_size(self):
        """
        sanitize_uploads() should flag a file larger than 400MB
        with True in the returned exclude_flag field.
        """
        self.utils = Utils()
        with tempfile.NamedTemporaryFile(suffix='.pdf') as temp:
            temp.write(b'0' * 410000000)
            tempname = temp.name
            files_dict_list = [{'name': tempname,
                                'full_path': tempname,
                                'exclude_flag': None,
                                'exclude_reason': None
                                }]
            docs = self.utils.sanitize_uploads(files_dict_list)
            self.assertTrue(docs[0]['exclude_flag'])

if __name__ == '__main__':
    unittest.main()


pneumatic/db.py
#!/usr/bin/env python

import os
import sys
import csv
import sqlite3
from colorama import init
from .utils import Utils


class Database(object):
    """
    A SQLite database to store results of file upload attempts to
    DocumentCloud, along with database-related utilities.
    """

    def __init__(self):
        self.utils = Utils()

        # Initialize colorama
        init()

    def make_db(self):
        # Create a SQLite db whose name includes a timestamp.
        timestamp = self.utils.timestamp()
        self.db_name = 'dc-upload-' + timestamp + '.db'

        # make the database file directory
        if not os.path.isdir('pneumatic_db'):
            os.mkdir('pneumatic_db')
        else:
            pass
        self.db_full_path = os.path.join('pneumatic_db', self.db_name)

        # Connect to the db and create the uploads table.
        if not os.path.exists(self.db_full_path):
            conn = sqlite3.connect(self.db_full_path)
            cur = conn.cursor()
            cur.execute('''
                CREATE TABLE uploads
                (
                    id Text, title Text, file_name Text, full_path Text,
                    upload_time Text, pages Integer, file_hash Text,
                    result Text, canonical_url Text, pdf_url Text,
                    text_url Text, exclude_flag Text, exclude_reason Text,
                    error_msg Text
                )
                       ''')
            conn.commit()
            conn.close()
            print('\033[36m* New uploads database created at: ' +
                  self.db_full_path)
        else:
            pass

    def insert_row(self, id, title, file_name, full_path, up_time, pages,
                   file_hash, result, canonical_url, pdf_url, text_url,
                   exclude_flag, exclude_reason, error_msg):
        """
        Inserts a row in the table.
        """
        conn = sqlite3.connect(self.db_full_path)
        cur = conn.cursor()
        cur.execute('''
            INSERT INTO uploads VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?);
            ''', (id, title, file_name, full_path, up_time, pages,
                  file_hash, result, canonical_url, pdf_url, text_url,
                  exclude_flag, exclude_reason, error_msg))
        conn.commit()
        conn.close()

    def update_row(self, db, id, title, file_name, full_path, up_time, pages,
                   file_hash, result, canonical_url, pdf_url, text_url,
                   exclude_flag, exclude_reason, error_msg):
        """
        Updates a row in the table.
        """
        conn = sqlite3.connect(db)
        cur = conn.cursor()

        cur.execute('''
            UPDATE uploads
            SET title = ?, pages = ?, file_hash = ?
            WHERE id = ?;
            ''', (title, pages, file_hash, id))
        conn.commit()
        conn.close()

    def dump_to_csv(self, db_name=None):
        """
        Outputs the contents of a SQLite database to CSV.
        """
        print('\n\033[36mDump to CSV')

        timestamp = self.utils.timestamp()

        # We can pass in a db name or use the one in the current session.
        # Check if it exists, first.
        if db_name:
            if os.path.isfile(db_name):
                db = db_name
            else:
                print('\033[31mERROR: \033[0mThe database file ' +
                      'specified does not exist.')
                sys.exit()
        else:
            db = self.db_full_path

        # Create an output folder and CSV file name.
        if not os.path.isdir('pneumatic_csv'):
            os.mkdir('pneumatic_csv')
        else:
            pass
        self.csv_name = 'dc-output-' + timestamp + '.csv'
        self.csv_full_path = os.path.join('pneumatic_csv', self.csv_name)

        # Query the database and write the rows to the CSV.
        row_counter = 0
        with open(self.csv_full_path, 'w', newline='') as csvfile:
            header_row = ('id', 'title', 'file_name', 'full_path',
                          'upload_time', 'pages', 'file_hash', 'result',
                          'canonical_url', 'pdf_url', 'text_url',
                          'exclude_flag', 'exclude_reason', 'error_msg')

            writer = csv.writer(csvfile, delimiter=',', quotechar='"')
            writer.writerow(header_row)

            # Query the database and write rows to CSV.
            conn = sqlite3.connect(db)
            cur = conn.cursor()
            for row in cur.execute('SELECT * FROM uploads;'):
                writer.writerow(row)
                row_counter += 1
            conn.close()

        print('\033[36m* CSV file created. ' + str(row_counter) +
              ' database records were exported to ' + self.csv_full_path)

    def cleanup_empty_db(self, db_name):
        """
        Checks for an empty db and removes if so.
        """
        conn = sqlite3.connect(db_name)
        cur = conn.cursor()
        cur.execute('SELECT COUNT(*) from uploads;')
        record_count = cur.fetchone()[0]
        conn.close()

        if record_count == 0:
            print('\n\033[36mCleanup\n* The new uploads database contains ' +
                  'no records. Deleting it to reduce clutter.')
            os.remove(db_name)
        else:
            pass

    def print_db_name(self):
        """
        Prints name of the database.
        """
        print('\n\033[36mDatabase Name\n* Responses from the DocumentCloud ' +
              'API are stored in a SQLite database in your current ' +
              'directory at: ' + self.db_full_path)


pneumatic/utils.py
#!/usr/bin/env python

import os
import time
from .excludes import file_excludes


class Utils(object):
    """
    A few things we'll (eventually) use.
    """

    def __init__(self):
        self.file_excludes = file_excludes

    def file_directory_check(self, file_directory):
        dir_exists = os.path.isdir(file_directory)
        return dir_exists

    def sanitize_uploads(self, doc_list):
        """
        Flag prohibited file types and files over 400MB upload size limit.
        """
        for x in doc_list:
            file_split = x['name'].split('.')
            if file_split[-1].lower() in self.file_excludes:
                x['exclude_flag'] = True
                x['exclude_reason'] = 'Prohibited file type'
            elif os.path.getsize(x['full_path']) > 400000000:
                x['exclude_flag'] = True
                x['exclude_reason'] = 'File over 400MB'
            else:
                x['exclude_flag'] = False
        return doc_list

    def timestamp(self):
        self.time = time.strftime("%Y%m%dT%H%M%S")
        return self.time


pneumatic/uploader.py
#!/usr/bin/env python

import os
import sys
import atexit
import json
import sqlite3
from multiprocessing import Pool
import requests
from colorama import init
from .db import Database
from .utils import Utils


class DocumentCloudUploader(object):
    """
    Main class containing functions to build and confirm file lists for
    uploading, make the upload requests, and later add metadata post-upload
    if desired.
    """

    def __init__(self, username, password):
        self.username = username
        self.password = password
        self.base_uri = 'https://www.documentcloud.org/api/'
        self.headers = {
            'User-Agent': 'pneumatic/0.1.8',
            'From': self.username
        }

        # Initialize colorama
        init()

        # Welcome users and test credentials
        print('\n\033[36mWelcome to pneumatic, a bulk uploader for ' +
              'DocumentCloud (https://www.documentcloud.org)\n\n' +
              'Initializing ...')
        if self.credential_test() == 200:
            print('\033[36m* Login credentials confirmed')
        else:
            print('\033[31mERROR: \033[0mYour username or password seems ' +
                  'to be incorrect.\nPlease check them and try again.\n\n' +
                  '\033[36mExiting pneumatic.\033[0m\n')
            sys.exit()

        # Initialize helper classes
        self.db = Database()
        self.utils = Utils()

        # Make the database that holds upload file data
        self.db.make_db()

        # Tasks to execute on program termination
        def cleanup(db_name):
            self.db.cleanup_empty_db(db_name)
            print('\n\033[36mExiting pneumatic.\033[0m\n')

        atexit.register(cleanup, self.db.db_full_path)

    def credential_test(self):
        """
        Confirm login credentials using simple search API endpoint.
        """
        print('\033[36m* Confirming your DocumentCloud login credentials')
        r = requests.get(self.base_uri + 'search.json?q=test', auth=(self.username, self.password))
        return r.status_code

    def log_exclusion(self, name, full_path, exclude_reason):
        """
        Log to the database a file excluded due to invalid file type
        or exceeding upload size limits.
        """
        timestamp = self.utils.timestamp()
        self.db.insert_row(
            None,            # id
            None,            # title
            name,
            full_path,
            timestamp,
            None,            # pages
            None,            # file hash
            None,            # result
            None,            # canonical_url
            None,            # pdf_url
            None,            # text_url
            'Y',             # exclude flag
            exclude_reason,
            None)            # error_msg

    def build_file_list(self, file_directory):
        """
        Walk the supplied or current directory, including sub-directories,
        and build a list of files for the upload. Report a file count to
        the user and give an option to proceed or not.
        """

        documents = []
        docs_to_upload = 0
        docs_to_exclude = 0

        print('\033[36m* Building a list of files to upload\n')

        # If no directory supplied, use the current directory.
        # If directory supplied, check that it exists.
        if not file_directory:
            file_directory = '.'
        else:
            directory_exists = self.utils.file_directory_check(file_directory)
            if not directory_exists:
                print('\n\033[31mERROR: \033[0mThe file directory ' +
                      'you supplied does not exist.\n' +
                      'Please check it and try again.')
                sys.exit()

        # Create a list of dictionaries that includes file name and full path,
        # plus placeholders for exclusion information.
        files_dict_list = []
        for root, dir, files in os.walk(file_directory):
            for f in files:
                file_dict = {}
                file_dict['name'] = f
                file_dict['full_path'] = os.path.join(root, f)
                file_dict['exclude_flag'] = None
                file_dict['exclude_reason'] = None
                files_dict_list.append(file_dict)
        #print(files_dict_list) # for testing for now

        # Check list of files for prohibited formats, size
        documents = self.utils.sanitize_uploads(files_dict_list)

        # Count documents to upload and exclude.
        for doc in documents:
            if doc['exclude_flag']:
                docs_to_exclude += 1
            else:
                docs_to_upload += 1
        print('pneumatic found ' + str(docs_to_upload) +
              ' file(s) to upload. ' +
              str(docs_to_exclude) + ' file(s) will be excluded.')

        # Ask user to continue.
        response = input('Begin upload? Y/n: ')
        if response.lower() == 'y':
            pass
        else:
            sys.exit()

        return documents

    def request(self, upload_dict):
        """
        Post the request and log response to the database.
        """
        files = {'file': open(upload_dict['full_path'], 'rb')}
        print('\033[0m.. Uploading ' + upload_dict['full_path'])
        timestamp = self.utils.timestamp()

        # Upload via requests.
        r = requests.post(self.base_uri + 'upload.json',
                          params=upload_dict['payload'],
                          files=files,
                          headers=self.headers,
                          auth=(self.username, self.password))

        # Check for success and set variables accordingly.
        if r.status_code == 200:
            upload_response = json.loads(r.text)
            print('\033[36m' + '++ Upload succeeded for ' +
                  upload_dict['full_path'])
            id = upload_response['id']
            title = upload_response['title']
            file_name = upload_dict['name']
            pages = upload_response['pages']
            file_hash = upload_response['file_hash']
            canonical_url = upload_response['canonical_url']
            pdf = upload_response['resources']['pdf']
            text = upload_response['resources']['text']
            error_msg = None
        else:
            print('\033[31m' + '!! Upload failed for ' +
                  upload_dict['full_path'])
            id = None
            title = None
            file_name = upload_dict['name']
            canonical_url = None
            pages = 0
            file_hash = None
            pdf = None
            text = None
            error_msg = r.text

        # Log upload response to the database.
        self.db.insert_row(
            id,
            title,
            file_name,
            upload_dict['full_path'],
            timestamp,
            pages,
            file_hash,
            r.status_code,
            canonical_url,
            pdf,
            text,
            'N',           # exclude_flag
            None,          # exclude_reason
            error_msg)     # error_msg

    def upload(self, file_directory=None, title=None, source=None,
               description=None, language=None, related_article=None,
               published_url=None, access='private', project=None,
               data=None, secure=False, force_ocr=False):
        """
        Upload one or more documents with associated metadata and options.
        """
        # Start with a list of files in the supplied directory. Each
        # file is a dict.
        documents = self.build_file_list(file_directory)

        # Appropriately process each file. Prohibited files will be
        # excluded and logged. The rest get added data and in the
        # list to be uploaded.
        cleared_uploads = []

        for doc in documents:
            if doc['exclude_flag']:
                self.log_exclusion(
                    doc['name'],
                    doc['full_path'],
                    doc['exclude_reason'])
            else:
                if title:
                    upload_title = title
                else:
                    upload_title = os.path.splitext(doc['name'])[0]

                doc['payload'] = {
                    'title': upload_title,
                    'source': source,
                    'description': description,
                    'language': language,
                    'related_article': related_article,
                    'published_url': published_url,
                    'access': access,
                    'project': project,
                }
                # Thank you to Ben Welsh (@palewire):
                if data:
                    for key, value in data.items():
                        doc['payload']['data[%s]' % key] = value

                if secure:
                    if secure is True:
                        doc['payload']['secure'] = 'true'

                if force_ocr:
                    if force_ocr is True:
                        doc['payload']['force_ocr'] = 'true'

                # print(doc['payload'])
                cleared_uploads.append(doc)

        # # Support multiprocessing on Linux-based systems
        # # Note: This is broken as of Python 3.8 and needs a refactor
        # # Multiprocessing in this context is not very useful anyway in hindsight

        # if os.name == 'posix':
        #     # Create pool of workers. Just 4 for now, please.
        #     p = Pool(processes=4)
        #     p.map(self.request, cleared_uploads)
        # # Otherwise, upload sequentially
        # elif os.name == 'nt':
        #     for upload_dict in cleared_uploads:
        #         self.request(upload_dict)

        for upload_dict in cleared_uploads:
            self.request(upload_dict)

    def request_get(self, db, url):
        """
        Request a document and update the database with results
        """

        r = requests.get(url, headers=self.headers, auth=(self.username, self.password))

        if r.status_code == 200:
            # update db
            get_response = json.loads(r.text)

            id = get_response['document']['id']
            title = get_response['document']['title']
            pages = get_response['document']['pages']
            file_hash = get_response['document']['file_hash']

            self.db.update_row(
                db,
                id,
                title,
                None,          # file name
                None,          # file path
                None,          # timestamp
                pages,
                file_hash,
                None,          # status_code
                None,          # canonical_url
                None,          # pdf_url
                None,          # text_url
                None,          # exclude_flag
                None,          # exclude_reason
                None)          # error_msg

        else:
            print('\033[37mDocument not found\033[0m')

    def update_processed_files(self, db_name=None):
        """
        Retrieve each file's page count and file hash once
        DocumentCloud has finished processing the files.
        """

        # We can pass in a db name or use the one in the current session.
        # Check if it exists, first.
        if db_name:
            if os.path.isfile(db_name):
                db = db_name
            else:
                print('\033[31mERROR: \033[0mThe database file ' +
                      'specified does not exist.')
                sys.exit()
        else:
            db = self.db.db_full_path

        print('\n\033[36mUpdate Processed Files\n* Updating documents in ' + db + ' with page count ' +
              'and file hash data.')

        # Generate list of API get requests for successfully uploaded document
        conn = sqlite3.connect(db)
        cur = conn.cursor()
        uploaded_doc_urls = [self.base_uri + 'documents/' + row[0] + '.json'
                             for row in cur.execute('SELECT id FROM uploads WHERE result = 200;')]
        conn.close()

        for url in uploaded_doc_urls:
            self.request_get(db, url)
